### ISSUE TITLE
[Epic #306] Phase 3: Nature-DQN-scale CNN policy (actor-critic + Q-network variants)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#327]
**Epic**: #306 — Atari (ALE) environment + CNN policy
**Phase**: 3 of 4
**Phase Goal**: A Burn conv policy at Nature-DQN scale, ready for the Atari workload.

## Description

Implement the Nature-DQN architecture (Mnih et al. 2015) as a Burn conv policy:

- Conv 32×8×8 stride 4 → Conv 64×4×4 stride 2 → Conv 64×3×3 stride 1 → FC 512 → heads
- Both actor-critic and Q-network head variants
- Prior art: `SnakeCnnBurnPolicy` (existing in-tree CNN policy) — follow its structure and conventions
- Seeded init consistent with `seeded_init`

Note: this issue does NOT depend on the ALE binding — the policy is pure Burn and can be built/tested on synthetic 84×84×4 tensors. It can proceed in parallel with Phase 2.

**Scope note**: subject to amendment by the Phase 1 design note (#324).

## Acceptance Criteria

- [ ] CNN policy with the Nature-DQN conv stack, actor-critic and Q-network variants, unit-tested on synthetic 84×84×4 input (output shapes, seeded-init determinism)
- [ ] Init behavior consistent with `seeded_init` conventions
- [ ] Tests + clippy -D warnings + fmt + doc build clean

## Dependencies

Part of Epic #306. Requires the Phase 1 spike (#324) verdict for scope confirmation only; technically independent of #325/#326 (parallelizable).

---
*Decomposed from Epic #306*

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating per explicit user request. All code references verified against `origin/main`.

---

### Prior Art in Detail

**`SnakeCnnBurnPolicy` — `src/policy/snake_cnn.rs`**

Module layout: `conv1/conv2/conv3: Conv2d<B>`, `fc_common: Linear<B>`, `fc_policy: Linear<B>`, `fc_value: Linear<B>`, `flat_size: usize`. No config struct — plain `new(grid_size, input_channels, device)`. Conv layers use `Conv2dConfig::new([in_ch, out_ch], [kH, kW]).with_padding(PaddingConfig2d::Same).init(device)`; FC layers use `linear_with_init` with `Initializer::KaimingUniform { gain: 1/√3, fan_out_only: false }`. Init is **unseeded** (Burn default). `flat_size` is cached as a plain module field so it survives `Module::load_record` — copy this pattern exactly. Input convention: NCHW `[B, C, H, W]`. Forward signature: `forward(grid: Tensor<B,4>) -> (logits [B, A], values [B, 1])` — value stays rank-2 `[B, 1]`; the trainer closure squeezes it (see snake example). Tests: construction + forward-shape checks at two batch sizes, all using `type B = Autodiff<NdArray<f32>>`.

**`MlpBurnPolicy` — `src/policy/mlp.rs` (actor-critic model)**

Has `MlpBurnConfig { hidden_dim, num_layers, use_orthogonal_init, activation, seed: Option<u64> }` with a `.with_seed(u64) -> Self` builder. PPO-facing method: `evaluate_actions(obs: Tensor<B,2>, actions: Tensor<B,1,Int>) -> (log_probs [B], entropy [B], values [B])` — all rank-1. `forward` returns `(logits [B, A], value [B])` — value is squeezed inside forward (rank-1 out). Seeded init is driven by `derive_layer_seed(base_seed, layer_idx)` + `seeded_layer_weights(seed, d_in, d_out, use_orthogonal, is_head)` + `linear_from_weights::<B>(d_in, d_out, &weights, device)`. Layer indices are fixed constants: `0=fc1, 1=fc2, 2=fc3, 3=policy_head, 4=value_head`. All three helpers are `pub(crate)` in `mlp.rs` and already imported by `q_network.rs`.

**`QNetworkBurn` — `src/policy/q_network.rs` (DQN model)**

Has `QNetworkBurnConfig { hidden_dim, use_orthogonal_init, seed: Option<u64> }`. Forward: `forward(obs: Tensor<B,2>) -> Q-values [B, A]`. Target-net sync: `copy_params_from(self, source: &QNetworkBurn<B>) -> QNetworkBurn<B>` (Burn record-based clone, returns new module). Seeded via the same three `mlp.rs` helpers; layer indices `0=fc1, 1=fc2, 2=q_head`.

---

### Trainer Interface — What the New Structs Must Satisfy

Both trainers are **closure-based**, not trait-based. The only module bounds are `AutodiffModule<B> + Clone`, automatically satisfied by `#[derive(Module, Debug)]`.

**`PPOTrainerBurn`** takes:
```rust
evaluate_fn: FnMut(&P, Tensor<B, 2>, Tensor<B, 1, Int>) -> (Tensor<B, 1>, Tensor<B, 1>, Tensor<B, 1>)
// obs is FLAT [B, C*H*W] from rollout buffer; closure must reshape to [B, C, H, W] before calling forward
// returns (action_log_probs [B], entropy_per_row [B], values [B])
```

See `examples/games/snake/train_snake_multi_v2.rs` lines 239–253 for the exact reshape closure pattern: `o_flat.reshape([b, NUM_CHANNELS, GRID_H, GRID_W])`, then `p.forward(o4)` returning `(logits, value)`, then inline log_softmax + entropy + action gather. The `NatureDqnBurnPolicy::evaluate_actions` method (taking rank-4 obs) should handle all of this internally — the closure just reshapes and delegates.

**`DQNTrainerBurn`** takes:
```rust
forward_fn: Fn(&Q, Tensor<B, 2>) -> Tensor<B, 2>   // obs FLAT → Q-values [B, A]
forward_target_fn: Fn(&Q, Tensor<B, 2>) -> Tensor<B, 2>
```
Same reshape pattern: closure reshapes flat obs to 4D before calling `q_net.forward`.

---

### `seeded_init` Conventions

Location: `src/policy/seeded_init.rs`. Two public functions:
- `seeded_orthogonal(seed, d_in, d_out, gain) -> Vec<f32>` — modified Gram–Schmidt QR, sign-fixed for cross-platform determinism
- `seeded_kaiming_uniform(seed, d_in, d_out, gain) -> Vec<f32>` — uniform in `[-bound, bound]`, `bound = gain × √(3 / fan_in)`

Consumption pattern (copy from `mlp.rs` / `q_network.rs`):
```rust
// derive_layer_seed, seeded_layer_weights, linear_from_weights are pub(crate) in mlp.rs
use super::mlp::{derive_layer_seed, linear_from_weights, linear_with_init, seeded_layer_weights};

if let Some(base_seed) = config.seed {
    let mut idx = 0u64;
    let mut next = || { let s = derive_layer_seed(base_seed, idx); idx += 1; s };
    let fc_common = { let w = seeded_layer_weights(next(), 3136, 512, false, false);
                      linear_from_weights::<B>(3136, 512, &w, device) };
    let policy_head = { let w = seeded_layer_weights(next(), 512, n_actions, false, true);
                        linear_from_weights::<B>(512, n_actions, &w, device) };
    // ... etc
}
```

**Conv layers are intentionally unseeded** (Burn's `Conv2dConfig` has no seed parameter, same gap as `LinearConfig`). The conv parameters total ~78K vs ~1.6M for `fc_common` alone, so this is a second-order concern. Document this decision in the module doc comment. The seeded path covers only the three FC layers.

Seeded FC layer indices to use (fixed, document in code):
- `NatureDqnBurnPolicy`: `0 = fc_common`, `1 = policy_head`, `2 = value_head`
- `NatureDqnQNetwork`: `0 = fc_common`, `1 = q_head`

---

### Exact Input Contract

- Layout: **NCHW** `[batch, channels, height, width]` — consistent with `SnakeCnnBurnPolicy` and Burn's `Conv2d`.
- Channels: 4 (frame-stack dimension; built by the preprocessor, not the policy).
- Spatial: 84 × 84.
- Dtype: `f32`. Pixel scaling convention: **0.0–1.0** (divide uint8 by 255). The policy is scale-agnostic but the doc comment should state the expectation.
- No batch-size constraint.

**Spatial dimensions after each conv (no padding, as per Nature-DQN spec):**

| Layer | Input | Kernel | Stride | Output |
|-------|-------|--------|--------|--------|
| conv1 | 84×84 | 8×8    | 4      | 20×20  |
| conv2 | 20×20 | 4×4    | 2      | 9×9    |
| conv3 | 9×9   | 3×3    | 1      | 7×7    |

Flat size: 64 × 7 × 7 = **3,136**

Burn conv config for conv1 (no padding, stride 4):
```rust
Conv2dConfig::new([4, 32], [8, 8])
    .with_stride([4, 4])
    .init(device)
```
Default `PaddingConfig2d` is `Valid` (no padding) — do NOT add `.with_padding(...)`.

---

### Parameter Count (4 actions, verified arithmetic)

| Layer             | Weights                    | Bias | Total     |
|-------------------|----------------------------|------|-----------|
| conv1 (32×4×8×8)  | 8,192                      | 32   | 8,224     |
| conv2 (64×32×4×4) | 32,768                     | 64   | 32,832    |
| conv3 (64×64×3×3) | 36,864                     | 64   | 36,928    |
| fc_common (3136→512) | 1,605,632               | 512  | 1,606,144 |
| policy_head (512→4)  | 2,048                   | 4    | 2,052     |
| value_head (512→1)   | 512                     | 1    | 513       |
| **NatureDqnBurnPolicy (AC)** |               |      | **1,686,693** |

For `NatureDqnQNetwork` (replace policy_head + value_head with q_head 512→4): **1,686,180**

The "~1.7M" figure cited in the issue is correct to 1 significant figure. The higher values sometimes quoted (~1.8M) assume 18 actions (full ALE). With 4 actions both variants are 1.686–1.687M.

**Builder sanity check**: after construction, iterate `module.collect_devices(...)` or manually sum `param.val().shape().num_elements()` over all `Linear` and `Conv2d` weights and biases. Expected: 1,686,693 (AC, n_actions=4) or 1,686,180 (Q-net, n_actions=4).

---

### Recommended Module Structure

**New file**: `src/policy/atari_cnn.rs`

Two structs (mirrors the MlpBurnPolicy / QNetworkBurn split):

```rust
// Config struct (following QNetworkBurnConfig pattern)
#[derive(Debug, Clone, Copy)]
pub struct NatureDqnConfig {
    pub seed: Option<u64>,
}
impl NatureDqnConfig {
    pub fn with_seed(mut self, seed: u64) -> Self { self.seed = Some(seed); self }
}

// Actor-critic variant
#[derive(Module, Debug)]
pub struct NatureDqnBurnPolicy<B: Backend> {
    conv1: Conv2d<B>, conv2: Conv2d<B>, conv3: Conv2d<B>,
    fc_common: Linear<B>,
    policy_head: Linear<B>,
    value_head: Linear<B>,
    flat_size: usize,    // = 3136; cached, survives load_record
}

// Q-network variant
#[derive(Module, Debug)]
pub struct NatureDqnQNetwork<B: Backend> {
    conv1: Conv2d<B>, conv2: Conv2d<B>, conv3: Conv2d<B>,
    fc_common: Linear<B>,
    q_head: Linear<B>,
    flat_size: usize,
}
```

Method surfaces:

`NatureDqnBurnPolicy`:
- `new(n_actions: usize, device: &B::Device) -> Self` — unseeded convenience ctor
- `with_config(n_actions: usize, config: NatureDqnConfig, device: &B::Device) -> Self` — seeded path
- `forward(&self, obs: Tensor<B,4>) -> (Tensor<B,2>, Tensor<B,2>)` — `(logits [B, A], values [B, 1])` — value retains rank-2 (same as snake_cnn)
- `evaluate_actions(&self, obs: Tensor<B,4>, actions: Tensor<B,1,Int>) -> (Tensor<B,1>, Tensor<B,1>, Tensor<B,1>)` — `(action_log_probs [B], entropy [B], values [B])` — mirrors `MlpBurnPolicy::evaluate_actions`

`NatureDqnQNetwork`:
- `new(n_actions: usize, device: &B::Device) -> Self`
- `with_config(n_actions: usize, config: NatureDqnConfig, device: &B::Device) -> Self`
- `forward(&self, obs: Tensor<B,4>) -> Tensor<B,2>` — `[B, n_actions]`
- `copy_params_from(self, source: &Self) -> Self` — mirrors `QNetworkBurn::copy_params_from`

---

### Acceptance Criteria (concrete, testable)

- [ ] `NatureDqnBurnPolicy::new(4, device).forward(Tensor::zeros([1, 4, 84, 84], device))` → shapes `([1, 4], [1, 1])`
- [ ] Same with batch=32: shapes `([32, 4], [32, 1])`
- [ ] `NatureDqnQNetwork::new(4, device).forward(Tensor::zeros([1, 4, 84, 84], device))` → shape `[1, 4]`
- [ ] `evaluate_actions` with batch=8 → all three outputs shape `[8]`
- [ ] Seeded init: two `with_config(4, NatureDqnConfig::default().with_seed(42), device)` constructions yield bit-identical FC weights for `fc_common`, `policy_head`, `value_head`; different seed differs (conv weights need not be tested for seeded identity)
- [ ] `derive_layer_seed` decorrelation: `fc_common.weight ≠ policy_head.weight` within one seeded construction
- [ ] `copy_params_from`: after sync, `q_net.forward(obs) == source.forward(obs)` element-wise
- [ ] Parameter count: sum of all param elements == **1,686,693** (AC, n_actions=4) and **1,686,180** (Q-net, n_actions=4)
- [ ] `cargo test --features training -- atari_cnn` passes
- [ ] `cargo clippy --features training -- -D warnings` clean (no new warnings)
- [ ] `cargo fmt --check` clean
- [ ] `cargo doc --features training --no-deps` clean (no broken doc links)

---

### Scope Guard

- No changes to `PPOTrainerBurn` or `DQNTrainerBurn`
- No env dependency, no ALE binding, no `env-atari` feature — that is #325/#326
- No benches — that is #328
- No changes to `SnakeCnnBurnPolicy`, `MlpBurnPolicy`, `QNetworkBurn`, or `seeded_init.rs`
- No changes to `Cargo.toml` (no new dependencies; the new file uses only existing Burn + seeded_init imports)
- Deliverable is exactly two file changes: one new file + one line added to `mod.rs`

---

## Affected Files

- `src/policy/atari_cnn.rs` — **new file** to create; primary deliverable
- `src/policy/mod.rs` — add `#[cfg(feature = "training")] pub mod atari_cnn;` (one line, after the `snake_cnn` entry at line 51)

## Test Plan

- [ ] `test_nature_dqn_ac_forward_single` — AC forward on `[1, 4, 84, 84]` → shapes `([1, 4], [1, 1])`
- [ ] `test_nature_dqn_ac_forward_batch` — AC forward on `[32, 4, 84, 84]` → shapes `([32, 4], [32, 1])`
- [ ] `test_nature_dqn_q_forward` — Q-net forward on `[1, 4, 84, 84]` → shape `[1, 4]`
- [ ] `test_nature_dqn_evaluate_actions_shapes` — evaluate_actions with batch=8, 4 actions → all outputs `[8]`
- [ ] `test_nature_dqn_seeded_fc_identical` — same seed → bit-identical FC weights; different seed differs
- [ ] `test_nature_dqn_seeded_layers_decorrelated` — fc_common ≠ policy_head within single seeded construction
- [ ] `test_nature_dqn_q_copy_params_from` — forward outputs agree after copy
- [ ] `test_nature_dqn_param_count_ac` — sum of param elements == 1,686,693 for n_actions=4
- [ ] `test_nature_dqn_param_count_q` — sum of param elements == 1,686,180 for n_actions=4

All tests inline as `#[cfg(test)] mod tests { ... }` inside `src/policy/atari_cnn.rs`, using `type B = Autodiff<NdArray<f32>>`.
